cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -179,6 +328,70 @@ size_type inplace_bitmask_binop(Binop op,
   return d_counter.value(stream);
 }
 
+/**
+ * @brief Performs a segmented bitwise operation `op` across multiple bitmasks and writes the
+ * results in-place to destination masks.
+ *
+ * This function performs bitwise operations on segments of bitmasks defined by segment_offsets,
+ * writing the results directly to the specified destination masks.
+ *
+ * @param[out] dest_masks Device span of pointers to destination bitmasks where results will be
+ * written
+ * @param[in] dest_mask_size The size of each destination mask in bitmask words
+ * @param[in] masks Host span of pointers to source bitmasks to be operated on
+ * @param[in] masks_begin_bits The bit offsets from which each source mask is to be operated on
+ * @param[in] mask_size_bits The number of bits to be operated on in each mask
+ * @param[in] segment_offsets Host span of offsets defining the segments for the operation
+ * @param[in] stream CUDA stream used for device memory operations and kernel launches
+ * @param[in] mr Device memory resource used to allocate temporary device memory
+ * @return A device vector containing the counts of unset bits in the destination mask corresponding
+ * to each segment after the AND operation
+ */
+template <typename Binop>
+rmm::device_uvector<size_type> inplace_segmented_bitmask_binop(
+  Binop op,
+  device_span<bitmask_type*> dest_masks,
+  size_type dest_mask_size,
+  host_span<bitmask_type const* const> masks,
+  host_span<size_type const> masks_begin_bits,
+  size_type mask_size_bits,
+  host_span<size_type const> segment_offsets,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(
+    std::all_of(masks_begin_bits.begin(), masks_begin_bits.end(), [](auto b) { return b >= 0; }),
+    "Invalid range.");
+  CUDF_EXPECTS(mask_size_bits > 0, "Invalid bit range.");
+  CUDF_EXPECTS(std::all_of(masks.begin(), masks.end(), [](auto p) { return p != nullptr; }),
+               "Mask pointer cannot be null");
+
+  rmm::device_uvector<size_type> d_null_counts(segment_offsets.size() - 1, stream, mr);
+  auto temp_mr      = cudf::get_current_device_resource_ref();
+  auto d_masks      = cudf::detail::make_device_uvector_async(masks, stream, temp_mr);
+  auto d_begin_bits = cudf::detail::make_device_uvector_async(masks_begin_bits, stream, temp_mr);
+  auto d_segment_offsets =
+    cudf::detail::make_device_uvector_async(segment_offsets, stream, temp_mr);
+  stream.synchronize();
+
+  auto constexpr block_size = 256;
+  auto constexpr warps_per_block =
+    util::div_rounding_up_safe<int>(block_size, cudf::detail::warp_size);
+  auto const num_blocks =
+    util::div_rounding_up_safe<int>(segment_offsets.size() - 1, warps_per_block);
```

Comment:
Optional: Maybe we can use a heuristic here to set `block_size` in case we have too few segments.